src/webgpu/api/validation/encoding/encoder_open_state.spec.ts

```diff
@@ -0,0 +1,383 @@
+    const srcTexture = t.device.createTexture({
+      size: textureSize,
+      format: textureFormat,
+      usage: GPUTextureUsage.COPY_SRC,
+    });
+    const dstTexture = t.device.createTexture({
+      size: textureSize,
+      format: textureFormat,
+      usage: GPUTextureUsage.COPY_DST,
+    });
+
+    const encoder = t.device.createCommandEncoder();
+
+    if (finishBeforeCommand) encoder.finish();
+
+    t.expectValidationError(() => {
+      switch (command) {
+        case 'clearBuffer':
+          {
+            encoder.clearBuffer(dstBuffer, 0, 16);
+          }
+          break;
+        case 'copyBufferToBuffer':
+          {
+            encoder.copyBufferToBuffer(srcBuffer, 0, dstBuffer, 0, 0);
+          }
+          break;
+        case 'copyBufferToTexture':
+          {
+            encoder.copyBufferToTexture(
+              { buffer: srcBuffer },
+              { texture: dstTexture },
+              textureSize
+            );
+          }
+          break;
+        case 'copyTextureToBuffer':
+          {
+            encoder.copyTextureToBuffer(
+              { texture: srcTexture },
+              { buffer: dstBuffer },
+              textureSize
+            );
+          }
+          break;
+        case 'copyTextureToTexture':
+          {
+            encoder.copyTextureToTexture(
+              { texture: srcTexture },
+              { texture: dstTexture },
+              textureSize
+            );
+          }
+          break;
+        case 'insertDebugMarker':
+          {
+            encoder.insertDebugMarker('marker');
+          }
+          break;
+        case 'pushDebugGroup':
+        case 'popDebugGroup':
+          {
+            encoder.pushDebugGroup('initializeWithStoreOp');
+            encoder.popDebugGroup();
+          }
+          break;
+        case 'writeTimestamp':
+          {
+            const querySet = t.device.createQuerySet({ type: 'timestamp', count: 1 });
+            encoder.writeTimestamp(querySet, 0);
+          }
+          break;
+        case 'resolveQuerySet':
+          {
+            const querySet = t.device.createQuerySet({ type: 'occlusion', count: 1 });
+            encoder.resolveQuerySet(querySet, 0, 1, dstBuffer, 0);
+          }
+          break;
+        default:
+          unreachable();
+      }
+    }, finishBeforeCommand);
+
+    if (!finishBeforeCommand) encoder.finish();
+  });
+
+g.test('render_pass_commands')
+  .params(u =>
+    u
+      .combine('command', kRenderPassEncoderCommands)
+      .beginSubcases()
+      .combine('finishBeforeCommand', [false, true])
+  )
+  .fn(t => {
+    const { command, finishBeforeCommand } = t.params;
+
+    const querySet = t.device.createQuerySet({ type: 'occlusion', count: 1 });
+    const encoder = t.device.createCommandEncoder();
+    const renderPass = beginRenderPassWithQuerySet(t, encoder, querySet);
+
+    if (finishBeforeCommand) {
+      renderPass.end();
+      encoder.finish();
```

Comment:
This finish shouldn't be strictly necessary, commands should fail if the pass is ended even if the encoder is still open. But it's best to test both, which would be 3 cases: nothing before command, end before command, end+finish before command.

Reply:
Good point. I would like to test it in the following PR. I add a TODO for now.